File: src-tauri/src/services/llm.rs
use std::future::Future;
use std::pin::Pin;

use reqwest::header::{AUTHORIZATION, CONTENT_TYPE};
use reqwest::Client;
use serde_json::json;
use sqlx::{Pool, Sqlite};

use crate::db::models::{LlmProviderConfig, LlmProviderType};

const ANTHROPIC_VERSION: &str = "2023-06-01";

#[derive(Debug, thiserror::Error)]
pub enum LlmServiceError {
    #[error("{0}")]
    Config(String),
    #[error("{0}")]
    Request(String),
    #[error("{0}")]
    Response(String),
}

pub trait LlmProvider: Send + Sync {
    fn config(&self) -> &LlmProviderConfig;
    fn test_connection<'a>(
        &'a self,
    ) -> Pin<Box<dyn Future<Output = Result<(), LlmServiceError>> + Send + 'a>>;
    fn generate<'a>(
        &'a self,
        system_prompt: &'a str,
        user_prompt: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<String, LlmServiceError>> + Send + 'a>>;
}

pub async fn load_provider_configs(
    db: &Pool<Sqlite>,
) -> Result<Vec<LlmProviderConfig>, LlmServiceError> {
    let row: Option<(String,)> = sqlx::query_as("SELECT value FROM configs WHERE key = ?1")
        .bind("llm_providers")
        .fetch_optional(db)
        .await
        .map_err(|error| LlmServiceError::Config(format!("读取模型配置失败: {error}")))?;

    let Some((raw,)) = row else {
        return Ok(Vec::new());
    };

    serde_json::from_str::<Vec<LlmProviderConfig>>(&raw)
        .map_err(|error| LlmServiceError::Config(format!("模型配置格式无效: {error}")))
}

pub fn select_provider(
    providers: &[LlmProviderConfig],
    requested_id: Option<&str>,
) -> Result<LlmProviderConfig, LlmServiceError> {
    if providers.is_empty() {
        return Err(LlmServiceError::Config(
            "请先在设置中配置至少一个大语言模型提供商".to_string(),
        ));
    }

    if let Some(provider_id) = requested_id.filter(|value| !value.trim().is_empty()) {
        return providers
            .iter()
            .find(|provider| provider.id == provider_id)
            .cloned()
            .ok_or_else(|| {
                LlmServiceError::Config(format!("未找到指定的大语言模型配置: {provider_id}"))
            });
    }

    providers
        .iter()
        .find(|provider| provider.is_default)
        .or_else(|| providers.first())
        .cloned()
        .ok_or_else(|| LlmServiceError::Config("没有可用的大语言模型配置".to_string()))
}

pub fn build_provider(
    config: LlmProviderConfig,
) -> Result<Box<dyn LlmProvider>, LlmServiceError> {
    validate_provider_config(&config)?;

    match config.provider_type {
        LlmProviderType::OpenAiCompatible => Ok(Box::new(OpenAiCompatibleProvider::new(config))),
        LlmProviderType::Anthropic => Ok(Box::new(AnthropicProvider::new(config))),
    }
}

fn validate_provider_config(config: &LlmProviderConfig) -> Result<(), LlmServiceError> {
    if config.name.trim().is_empty() {
        return Err(LlmServiceError::Config("模型配置名称不能为空".to_string()));
    }
    if config.model.trim().is_empty() {
        return Err(LlmServiceError::Config("模型名称不能为空".to_string()));
    }
    if config.api_key.trim().is_empty() {
        return Err(LlmServiceError::Config("API Key 不能为空".to_string()));
    }

    Ok(())
}

struct OpenAiCompatibleProvider {
    client: Client,
    config: LlmProviderConfig,
}

impl OpenAiCompatibleProvider {
    fn new(config: LlmProviderConfig) -> Self {
        Self {
            client: Client::new(),
            config,
        }
    }

    fn base_url(&self) -> String {
        self.config
            .base_url
            .as_deref()
            .filter(|value| !value.trim().is_empty())
            .unwrap_or("https://api.openai.com/v1")
            .trim_end_matches('/')
            .to_string()
    }
}

impl LlmProvider for OpenAiCompatibleProvider {
    fn config(&self) -> &LlmProviderConfig {
        &self.config
    }

    fn test_connection<'a>(
        &'a self,
    ) -> Pin<Box<dyn Future<Output = Result<(), LlmServiceError>> + Send + 'a>> {
        Box::pin(async move {
            let response = self
                .client
                .post(format!("{}/chat/completions", self.base_url()))
                .header(AUTHORIZATION, format!("Bearer {}", self.config.api_key.trim()))
                .header(CONTENT_TYPE, "application/json")
                .json(&json!({
                    "model": self.config.model,
                    "max_tokens": 1,
                    "temperature": 0,
                    "messages": [
                        { "role": "user", "content": "ping" }
                    ]
                }))
                .send()
                .await
                .map_err(map_request_error)?;

            let json = ensure_success(response).await?;
            extract_openai_text(&json).map(|_| ())
        })
    }

    fn generate<'a>(
        &'a self,
        system_prompt: &'a str,
        user_prompt: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<String, LlmServiceError>> + Send + 'a>> {
        Box::pin(async move {
            let response = self
                .client
                .post(format!("{}/chat/completions", self.base_url()))
                .header(AUTHORIZATION, format!("Bearer {}", self.config.api_key.trim()))
                .header(CONTENT_TYPE, "application/json")
                .json(&json!({
                    "model": self.config.model,
                    "temperature": 0.2,
                    "messages": [
                        { "role": "system", "content": system_prompt },
                        { "role": "user", "content": user_prompt }
                    ]
                }))
                .send()
                .await
                .map_err(map_request_error)?;

            let json = ensure_success(response).await?;
            extract_openai_text(&json)
        })
    }
}

struct AnthropicProvider {
    client: Client,
    config: LlmProviderConfig,
}

impl AnthropicProvider {
    fn new(config: LlmProviderConfig) -> Self {
        Self {
            client: Client::new(),
            config,
        }
    }

    fn base_url(&self) -> String {
        self.config
            .base_url
            .as_deref()
            .filter(|value| !value.trim().is_empty())
            .unwrap_or("https://api.anthropic.com/v1")
            .trim_end_matches('/')
            .to_string()
    }
}

impl LlmProvider for AnthropicProvider {
    fn config(&self) -> &LlmProviderConfig {
        &self.config
    }

    fn test_connection<'a>(
        &'a self,
    ) -> Pin<Box<dyn Future<Output = Result<(), LlmServiceError>> + Send + 'a>> {
        Box::pin(async move {
            let response = self
                .client
                .post(format!("{}/messages", self.base_url()))
                .header("x-api-key", self.config.api_key.trim())
                .header("anthropic-version", ANTHROPIC_VERSION)
                .header(CONTENT_TYPE, "application/json")
                .json(&json!({
                    "model": self.config.model,
                    "max_tokens": 1,
                    "messages": [
                        {
                            "role": "user",
                            "content": "ping"
                        }
                    ]
                }))
                .send()
                .await
                .map_err(map_request_error)?;

            let json = ensure_success(response).await?;
            extract_anthropic_text(&json).map(|_| ())
        })
    }

    fn generate<'a>(
        &'a self,
        system_prompt: &'a str,
        user_prompt: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<String, LlmServiceError>> + Send + 'a>> {
        Box::pin(async move {
            let response = self
                .client
                .post(format!("{}/messages", self.base_url()))
                .header("x-api-key", self.config.api_key.trim())
                .header("anthropic-version", ANTHROPIC_VERSION)
                .header(CONTENT_TYPE, "application/json")
                .json(&json!({
                    "model": self.config.model,
                    "max_tokens": 1800,
                    "temperature": 0.2,
                    "system": system_prompt,
                    "messages": [
                        {
                            "role": "user",
                            "content": user_prompt
                        }
                    ]
                }))
                .send()
                .await
                .map_err(map_request_error)?;

            let json = ensure_success(response).await?;
            extract_anthropic_text(&json)
        })
    }
}

async fn ensure_success(
    response: reqwest::Response,
) -> Result<serde_json::Value, LlmServiceError> {
    let status = response.status();
    let body = response
        .text()
        .await
        .map_err(|error| LlmServiceError::Response(format!("读取模型响应失败: {error}")))?;

    if !status.is_success() {
        return Err(LlmServiceError::Request(format!(
            "模型请求失败（{}）: {}",
            status,
            truncate_message(&body)
        )));
    }

    serde_json::from_str::<serde_json::Value>(&body)
        .map_err(|error| LlmServiceError::Response(format!("解析模型响应失败: {error}")))
}

fn extract_openai_text(json: &serde_json::Value) -> Result<String, LlmServiceError> {
    let content = json
        .get("choices")
        .and_then(|choices| choices.as_array())
        .and_then(|choices| choices.first())
        .and_then(|choice| choice.get("message"))
        .and_then(|message| message.get("content"));

    extract_string_or_parts(content).ok_or_else(|| {
        LlmServiceError::Response("OpenAI 兼容模型未返回可用的文本内容".to_string())
    })
}

fn extract_anthropic_text(json: &serde_json::Value) -> Result<String, LlmServiceError> {
    let content = json.get("content");
    extract_string_or_parts(content).ok_or_else(|| {
        LlmServiceError::Response("Anthropic 模型未返回可用的文本内容".to_string())
    })
}

fn extract_string_or_parts(value: Option<&serde_json::Value>) -> Option<String> {
    match value {
        Some(serde_json::Value::String(text)) => {
            let trimmed = text.trim();
            (!trimmed.is_empty()).then(|| trimmed.to_string())
        }
        Some(serde_json::Value::Array(parts)) => {
            let text = parts
                .iter()
                .filter_map(|item| {
                    item.get("text")
                        .and_then(|value| value.as_str())
                        .map(str::trim)
                })
                .filter(|text| !text.is_empty())
                .collect::<Vec<_>>()
                .join("\n\n");

            (!text.trim().is_empty()).then(|| text)
        }
        _ => None,
    }
}

fn map_request_error(error: reqwest::Error) -> LlmServiceError {
    if error.is_timeout() {
        return LlmServiceError::Request("连接模型服务超时，请稍后重试".to_string());
    }

    LlmServiceError::Request(format!("连接模型服务失败: {error}"))
}

fn truncate_message(message: &str) -> String {
    const LIMIT: usize = 320;
    if message.len() <= LIMIT {
        return message.trim().to_string();
    }

    format!("{}...", message[..LIMIT].trim())
}

File: internal/llm/llm.go
package llm

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os/exec"
	"strings"
	"time"

	"github.com/blueship581/pinru/internal/util"
)

const anthropicVersion = "2023-06-01"

type Provider interface {
	Name() string
	Model() string
	TestConnection() error
	Generate(systemPrompt, userPrompt string) (string, error)
}

type Config struct {
	ID             string  `json:"id"`
	Name           string  `json:"name"`
	ProviderType   string  `json:"providerType"`
	Model          string  `json:"model"`
	BaseURL        *string `json:"baseUrl"`
	APIKey         string  `json:"apiKey"`
	IsDefault      bool    `json:"isDefault"`
	ThinkingBudget string  `json:"thinkingBudget"` // "" | "low" | "medium" | "high"
}

func IsACPProvider(providerType string) bool {
	return providerType == "claude_code_acp" || providerType == "codex_acp"
}

func thinkingBudgetTokens(budget string) int {
	switch budget {
	case "low":
		return 1024
	case "medium":
		return 4096
	case "high":
		return 10240
	default:
		return 0
	}
}

func BuildProvider(cfg Config) (Provider, error) {
	if !IsACPProvider(cfg.ProviderType) && strings.TrimSpace(cfg.APIKey) == "" {
		return nil, fmt.Errorf("API Key 不能为空")
	}
	switch cfg.ProviderType {
	case "openai_compatible":
		timeout := 60 * time.Second
		if cfg.ThinkingBudget != "" {
			timeout = 180 * time.Second
		}
		return &openaiProvider{cfg: cfg, client: &http.Client{Timeout: timeout}}, nil
	case "anthropic":
		timeout := 60 * time.Second
		if cfg.ThinkingBudget != "" {
			timeout = 180 * time.Second
		}
		return &anthropicProvider{cfg: cfg, client: &http.Client{Timeout: timeout}}, nil
	case "claude_code_acp":
		return &claudeCodeACPProvider{cfg: cfg}, nil
	case "codex_acp":
		return &codexACPProvider{cfg: cfg}, nil
	default:
		return nil, fmt.Errorf("unknown provider type: %s", cfg.ProviderType)
	}
}

// --- OpenAI Compatible Provider ---

type openaiProvider struct {
	cfg    Config
	client *http.Client
}

func (p *openaiProvider) Name() string  { return p.cfg.Name }
func (p *openaiProvider) Model() string { return p.cfg.Model }
func (p *openaiProvider) baseURL() string {
	if p.cfg.BaseURL != nil && strings.TrimSpace(*p.cfg.BaseURL) != "" {
		return strings.TrimRight(*p.cfg.BaseURL, "/")
	}
	return "https://api.openai.com/v1"
}

func (p *openaiProvider) TestConnection() error {
	body, _ := json.Marshal(map[string]interface{}{
		"model":       p.cfg.Model,
		"max_tokens":  1,
		"temperature": 0,
		"messages":    []map[string]string{{"role": "user", "content": "ping"}},
	})
	resp, err := p.doRequest(body)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	return checkLLMResponse(resp)
}

func (p *openaiProvider) Generate(systemPrompt, userPrompt string) (string, error) {
	payload := map[string]interface{}{
		"model":       p.cfg.Model,
		"temperature": 0.2,
		"messages": []map[string]string{
			{"role": "system", "content": systemPrompt},
			{"role": "user", "content": userPrompt},
		},
	}
	if p.cfg.ThinkingBudget != "" {
		payload["reasoning_effort"] = p.cfg.ThinkingBudget
	}
	body, _ := json.Marshal(payload)
	resp, err := p.doRequest(body)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if err := checkLLMResponse(resp); err != nil {
		return "", err
	}
	return extractOpenAIText(resp)
}

func (p *openaiProvider) doRequest(body []byte) (*http.Response, error) {
	req, _ := http.NewRequest("POST", p.baseURL()+"/chat/completions", bytes.NewReader(body))
	req.Header.Set("Authorization", "Bearer "+strings.TrimSpace(p.cfg.APIKey))
	req.Header.Set("Content-Type", "application/json")
	return p.client.Do(req)
}

// --- Anthropic Provider ---

type anthropicProvider struct {
	cfg    Config
	client *http.Client
}

func (p *anthropicProvider) Name() string  { return p.cfg.Name }
func (p *anthropicProvider) Model() string { return p.cfg.Model }
func (p *anthropicProvider) baseURL() string {
	if p.cfg.BaseURL != nil && strings.TrimSpace(*p.cfg.BaseURL) != "" {
		return strings.TrimRight(*p.cfg.BaseURL, "/")
	}
	return "https://api.anthropic.com/v1"
}

func (p *anthropicProvider) TestConnection() error {
	body, _ := json.Marshal(map[string]interface{}{
		"model":      p.cfg.Model,
		"max_tokens": 1,
		"messages":   []map[string]string{{"role": "user", "content": "ping"}},
	})
	resp, err := p.doRequest(body)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	return checkLLMResponse(resp)
}

func (p *anthropicProvider) Generate(systemPrompt, userPrompt string) (string, error) {
	budgetTokens := thinkingBudgetTokens(p.cfg.ThinkingBudget)
	payload := map[string]interface{}{
		"model":  p.cfg.Model,
		"system": systemPrompt,
		"messages": []map[string]interface{}{
			{"role": "user", "content": userPrompt},
		},
	}
	if budgetTokens > 0 {
		payload["max_tokens"] = 16384
		payload["temperature"] = 1
		payload["thinking"] = map[string]interface{}{
			"type":          "enabled",
			"budget_tokens": budgetTokens,
		}
	} else {
		payload["max_tokens"] = 4096
		payload["temperature"] = 0.2
	}
	body, _ := json.Marshal(payload)
	resp, err := p.doRequest(body)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if err := checkLLMResponse(resp); err != nil {
		return "", err
	}
	return extractAnthropicText(resp)
}

func (p *anthropicProvider) doRequest(body []byte) (*http.Response, error) {
	req, _ := http.NewRequest("POST", p.baseURL()+"/messages", bytes.NewReader(body))
	req.Header.Set("x-api-key", strings.TrimSpace(p.cfg.APIKey))
	req.Header.Set("anthropic-version", anthropicVersion)
	req.Header.Set("Content-Type", "application/json")
	return p.client.Do(req)
}

// --- Helpers ---

func checkLLMResponse(resp *http.Response) error {
	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		return nil
	}
	body, _ := io.ReadAll(resp.Body)
	msg := string(body)
	if len(msg) > 320 {
		msg = msg[:320] + "..."
	}
	return fmt.Errorf("模型请求失败（%d）: %s", resp.StatusCode, strings.TrimSpace(msg))
}

func extractOpenAIText(resp *http.Response) (string, error) {
	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("解析模型响应失败: %w", err)
	}
	choices, _ := result["choices"].([]interface{})
	if len(choices) == 0 {
		return "", fmt.Errorf("OpenAI 兼容模型未返回可用的文本内容")
	}
	choice, _ := choices[0].(map[string]interface{})
	msg, _ := choice["message"].(map[string]interface{})
	content, _ := msg["content"].(string)
	if strings.TrimSpace(content) == "" {
		return "", fmt.Errorf("OpenAI 兼容模型未返回可用的文本内容")
	}
	return strings.TrimSpace(content), nil
}

func extractAnthropicText(resp *http.Response) (string, error) {
	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("解析模型响应失败: %w", err)
	}
	contentArr, _ := result["content"].([]interface{})
	var texts []string
	for _, item := range contentArr {
		block, _ := item.(map[string]interface{})
		if text, ok := block["text"].(string); ok && strings.TrimSpace(text) != "" {
			texts = append(texts, strings.TrimSpace(text))
		}
	}
	if len(texts) == 0 {
		return "", fmt.Errorf("Anthropic 模型未返回可用的文本内容")
	}
	return strings.Join(texts, "\n\n"), nil
}

// --- Claude Code ACP Provider ---

type claudeCodeACPProvider struct {
	cfg Config
}

func (p *claudeCodeACPProvider) Name() string  { return p.cfg.Name }
func (p *claudeCodeACPProvider) Model() string { return p.cfg.Model }

func (p *claudeCodeACPProvider) TestConnection() error {
	// 先检查 CLI 是否安装（支持打包应用的受限 PATH 环境）
	claudePath, err := util.ResolveCLI("claude")
	if err != nil {
		return fmt.Errorf("Claude Code CLI 未安装，请执行: npm install -g @anthropic-ai/claude-code")
	}

	// 真实发一次 API 请求，验证 ACP 代理可达且有可用账号
	// 不传 --model，让全局配置决定模型，与提示词生成保持一致
	args := []string{"-p", "respond with the single word: ok", "--dangerously-skip-permissions"}
	cmd := exec.Command(claudePath, args...)

	type result struct {
		out []byte
		err error
	}
	ch := make(chan result, 1)
	go func() {
		out, err := cmd.CombinedOutput()
		ch <- result{out, err}
	}()

	timer := time.NewTimer(60 * time.Second)
	defer timer.Stop()
	select {
	case <-timer.C:
		_ = cmd.Process.Kill()
		return fmt.Errorf("Claude Code ACP 连接超时（60s），请检查网络或 ACP 配置")
	case r := <-ch:
		output := strings.TrimSpace(string(r.out))
		if r.err != nil {
			if strings.Contains(output, "No available accounts") || strings.Contains(output, "no available accounts") {
				return fmt.Errorf("ACP 账号池暂时耗尽（503），请稍后重试")
			}
			if strings.Contains(output, "模型配置不存在") {
				return fmt.Errorf("ACP 代理不支持当前模型，请检查 ACP 配置")
			}
			if output != "" {
				return fmt.Errorf("Claude Code ACP 调用失败: %s", output)
			}
			return fmt.Errorf("Claude Code ACP 调用失败: %v", r.err)
		}
		return nil
	}
}

func (p *claudeCodeACPProvider) Generate(systemPrompt, userPrompt string) (string, error) {
	claudePath, err := util.ResolveCLI("claude")
	if err != nil {
		return "", fmt.Errorf("Claude Code CLI 未安装，请执行: npm install -g @anthropic-ai/claude-code")
	}
	prompt := systemPrompt + "\n\n" + userPrompt
	args := []string{"--print", "--model", p.cfg.Model, prompt}
	cmd := exec.Command(claudePath, args...)
	cmd.Stdin = nil
	out, err := cmd.Output()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			return "", fmt.Errorf("Claude Code CLI 调用失败: %s", strings.TrimSpace(string(exitErr.Stderr)))
		}
		return "", fmt.Errorf("Claude Code CLI 调用失败: %w", err)
	}
	result := strings.TrimSpace(string(out))
	if result == "" {
		return "", fmt.Errorf("Claude Code CLI 未返回可用内容")
	}
	return result, nil
}

// --- Codex ACP Provider ---

type codexACPProvider struct {
	cfg Config
}

func (p *codexACPProvider) Name() string  { return p.cfg.Name }
func (p *codexACPProvider) Model() string { return p.cfg.Model }

func (p *codexACPProvider) TestConnection() error {
	codexPath, err := util.ResolveCLI("codex")
	if err != nil {
		return fmt.Errorf("Codex CLI 未安装或不可用: %v", err)
	}
	cmd := exec.Command(codexPath, "--version")
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("Codex CLI 未安装或不可用: %v", strings.TrimSpace(string(out)))
	}
	return nil
}

func (p *codexACPProvider) Generate(systemPrompt, userPrompt string) (string, error) {
	codexPath, err := util.ResolveCLI("codex")
	if err != nil {
		return "", fmt.Errorf("Codex CLI 未安装，请先安装后重试")
	}
	prompt := systemPrompt + "\n\n" + userPrompt
	args := []string{"--quiet", "--model", p.cfg.Model, prompt}
	cmd := exec.Command(codexPath, args...)
	out, err := cmd.Output()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			return "", fmt.Errorf("Codex CLI 调用失败: %s", strings.TrimSpace(string(exitErr.Stderr)))
		}
		return "", fmt.Errorf("Codex CLI 调用失败: %w", err)
	}
	result := strings.TrimSpace(string(out))
	if result == "" {
		return "", fmt.Errorf("Codex CLI 未返回可用内容")
	}
	return result, nil
}
